prow/servemux/servemux.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servemux
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/pluginhelp/externalplugins"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	eventTypeField = "event-type"
+)
+
+type EventHandler interface {
```

Comment:
I think the point Steve was making earlier is that we should not export an `EventHandler`, we should support a `plugin.Registry` because the `EventHandler` is just an implementation detail that doesn't really matter for plugin authors:
```
type PluginRegistry interface {
  RegisterGenericCommentHandler(name string, fn GenericCommentHandler, help HelpProvider) error
  RegisterReviewCommentEventHandler(name string, fn ReviewCommentEventHandler, help HelpProvider) error
  ...
}

func NewGitHubEventHandler(listenAddress string) EventHandler {

}
```

Furthermore, I don't think `servermux` is a good name for the package. A `ServerMux` is a generic thing for any kind of http server, but what we are building here is not, its specific for handling GithUb events.

From a reviewer perspective I would btw prefer it if you first did only the new package and not adjust hook. Right now I have to skim over 300 LOC hook changes that are irrelevant until we agree on how this here looks like